File: mango/mango-platform/mango-authorization/mango-authorization-core/src/main/java/io/mango/authorization/core/entity/Role.java
package io.mango.authorization.core.entity;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import lombok.Data;

import java.io.Serializable;
import java.time.LocalDateTime;

/**
 * 角色实体。
 */
@Data
@TableName("authorization_role")
public class Role implements Serializable {

    private static final long serialVersionUID = 1L;

    /** 角色 ID。 */
    @TableId(value = "id", type = IdType.ASSIGN_ID)
    private Long roleId;

    /** 租户 ID。 */
    private Long tenantId;

    /** 应用编码。 */
    private String appCode;

    /** 登录域。 */
    private String realm;

    /** 操作者类型。 */
    private String actorType;

    /** 角色编码。 */
    private String roleCode;

    /** 角色名称。 */
    private String roleName;

    /** 角色类型：1-系统角色，2-业务角色。 */
    private Integer roleType;

    /** 状态：0-禁用，1-启用。 */
    private Integer status;

    /** 排序号。 */
    private Integer sort;

    /** 创建时间。 */
    private LocalDateTime createTime;

    /** 更新时间。 */
    private LocalDateTime updateTime;

    /** 备注。 */
    private String remark;
}

File: mango-ui/packages/rbac/src/api/role.ts
import { del, get, post, put } from '@mango/common/utils/request';
import type { ApiId } from '@mango/api-schema';
import type { SysMenuVO } from './menu';

export interface RoleVO {
  roleId?: ApiId;
  tenantId?: ApiId;
  appCode: string;
  realm: string;
  actorType?: string;
  roleCode: string;
  roleName: string;
  roleType: number;
  status: number;
  sort?: number;
  remark?: string;
  createTime?: string;
  updateTime?: string;
  menuIds?: ApiId[];
  permissionIds?: ApiId[];
}

export type DataScopeMode = 'ALL' | 'SELF' | 'SELF_ORG' | 'SELF_ORG_AND_CHILDREN' | 'ORG';

export interface RoleDataScopeVO {
  id?: ApiId;
  tenantId?: ApiId;
  appCode?: string;
  roleId: ApiId;
  resourceCode: string;
  scopeMode: DataScopeMode;
  scopeValues?: string[];
  includeChildren?: boolean;
  status?: number;
  createTime?: string;
  updateTime?: string;
}

export interface SaveRoleDataScopeCommand {
  roleId: ApiId;
  resourceCode: string;
  scopeMode: DataScopeMode;
  scopeValues?: string[];
  includeChildren?: boolean;
  status?: number;
}

export const roleApi = {
  list: () => get<RoleVO[]>('/authorization/roles'),
  detail: (id: ApiId) => get<RoleVO>('/authorization/roles/detail', { params: { id } }),
  create: (data: RoleVO) => post<ApiId>('/authorization/roles', data),
  update: (data: RoleVO) => put<boolean>('/authorization/roles', data),
  delete: (id: ApiId) => del<boolean>('/authorization/roles', { params: { id } }),
  getMenuIds: (roleId: ApiId) => get<ApiId[]>('/authorization/roles/menus', { params: { roleId } }),
  getAssignableMenus: (appCode?: string) => get<SysMenuVO[]>('/authorization/roles/assignable-menus', { params: { appCode } }),
  assignMenus: (roleId: ApiId, menuIds: ApiId[]) => post<boolean>('/authorization/roles/menus', { roleId, menuIds }),
  getSubjectRoles: (subjectId: ApiId) => get<RoleVO[]>('/authorization/roles/subjects', { params: { subjectId } }),
  assignSubjectRoles: (data: {
    subjectId: ApiId;
    appCode?: string;
    realm?: string;
    actorType?: string;
    partyType?: string;
    partyId?: ApiId;
    roleIds: ApiId[];
  }) => post<boolean>('/authorization/roles/subjects', data),
  getDataScopes: (roleId: ApiId) => get<RoleDataScopeVO[]>('/authorization/data-scopes/roles', { params: { roleId } }),
  saveDataScope: (data: SaveRoleDataScopeCommand) => post<boolean>('/authorization/data-scopes/roles', data),
  deleteDataScope: (roleId: ApiId, resourceCode: string) => del<boolean>('/authorization/data-scopes/roles', {
    params: { roleId, resourceCode },
  }),
};
